### PR TITLE
generic way to send STS requests using mssql API

### PR DESCRIPTION
**Pull request by VasuBhog**
Add a way to send STS requests and get back the response. Will be used in SQL bindings extension to send addSqlBinding send request.

### CODE REVIEW

File: src/controllers/connectionManager.ts [hunk 1]
```diff
@@ -289,6 +289,18 @@ export default class ConnectionManager {
...
 	}
 
+	/**
+	 * Send a request to the SQL Tools Server client
+	 * @param type The of the request to make
+	 * @param params The params to pass with the request
+	 * @returns A thenable object for when the request receives a response
+	 */
+	public sendRequest<P, R, E, R0>(requestType: RequestType<P, R, E, R0>, params?: P): Thenable<R> {
+		if (this.client !== undefined) {
```
Comment (Charles-Gagnon) [inline src/controllers/connectionManager.ts:299]: `this.client?.sendRequest(...)`

But also - why is this necessary? I wouldn't expect client to ever be undefined. It's passed in or set in the constructor.

File: src/controllers/connectionManager.ts [hunk 2]
```diff
@@ -289,6 +289,18 @@ export default class ConnectionManager {
 		return ConnectionCredentials.createConnectionDetails(connectionInfo);
 	}
 
+	/**
+	 * Send a request to the SQL Tools Server client
+	 * @param type The of the request to make
+	 * @param params The params to pass with the request
+	 * @returns A thenable object for when the request receives a response
+	 */
+	public sendRequest<P, R, E, R0>(requestType: RequestType<P, R, E, R0>, params?: P): Thenable<R> {
```
Comment (Charles-Gagnon) [inline src/controllers/connectionManager.ts:298]: We generally shouldn't be using Thenables in our code - that's only something we do for extension APIs. This should be returning `Promise<R>`

File: src/controllers/connectionManager.ts [hunk 3]
```diff
@@ -289,6 +289,18 @@ export default class ConnectionManager {
 		return ConnectionCredentials.createConnectionDetails(connectionInfo);
 	}
 
+	/**
+	 * Send a request to the SQL Tools Server client
+	 * @param type The of the request to make
```
Comment (Charles-Gagnon) [inline src/controllers/connectionManager.ts:294]: requestType

File: src/controllers/connectionManager.ts (hunk 2 above)
Comment (VasuBhog) [reply src/controllers/connectionManager.ts:298]: The issue here is that the returned sendRequest of type Thenable ([serverclient.ts](https://github.com/microsoft/vscode-mssql/blob/main/src/languageservice/serviceclient.ts#L380)) which then also calls the sendRequest from vscode-languageclient. They are all of type Thenable.

Comment (Charles-Gagnon) [reply src/controllers/connectionManager.ts:298]: You can solve this by making this function async, awaiting the call to sendRequest and then returning the result

File: typings/vscode-mssql.d.ts [hunk 4]
```diff
@@ -3,9 +3,11 @@
  *  Licensed under the Source EULA. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+
```
Comment (chlafreniere) [inline typings/vscode-mssql.d.ts:6]: nit: extra space

File: src/controllers/connectionManager.ts [hunk 5]
```diff
@@ -289,6 +289,16 @@ export default class ConnectionManager {
 		return ConnectionCredentials.createConnectionDetails(connectionInfo);
 	}
 
+	/**
+	 * Send a request to the SQL Tools Server client
+	 * @param requestType The of the request to make
```
Comment (chlafreniere) [inline src/controllers/connectionManager.ts:294]: nit: "The type of the request"

Can we make this a bit clearer, by the way?

File: src/controllers/connectionManager.ts [hunk 6]
```diff
@@ -289,6 +289,16 @@ export default class ConnectionManager {
...
+	 * @param params The params to pass with the request
+	 * @returns A thenable object for when the request receives a response
```
Comment (chlafreniere) [inline src/controllers/connectionManager.ts:296]: promise not thenable

File: src/controllers/connectionManager.ts (hunk 5 above)
Comment (VasuBhog) [reply src/controllers/connectionManager.ts:294]: Not sure how much clearer since this is supposed to be generic to be used by many different contract request types

**Review by chlafreniere — APPROVED** [review]
